apollo-router-core/src/layers/apq.rs

```diff
@@ -0,0 +1,427 @@
+use crate::{plugin_utils::structures::RouterResponseBuilder, RouterRequest, RouterResponse};
+use futures::future::BoxFuture;
+use moka::sync::Cache;
+use serde::Deserialize;
+use serde_json_bytes::json;
+use sha2::{Digest, Sha256};
+use std::task::Poll;
+use tower::{BoxError, Layer, Service};
+
+#[derive(Deserialize, Clone, Debug)]
+pub struct PersistedQuery {
+    pub version: u8,
+    #[serde(rename = "sha256Hash")]
+    pub sha256hash: String,
+}
+
+#[derive(Clone)]
```

Comment:
Not sure that we need this #[derive]. It depends on how we address the requirement to not clone apq in the `call()` fn.

Reply:
the [cache we rely on](https://github.com/moka-rs/moka/blob/89736fa3aeb72d026a2783a3c642e3f661d0e2e5/src/sync/cache.rs?_pjax=%23repo-content-pjax-container#L228) derives(Clone) and 

```Cache<K, V, S>` requires trait bounds `Send`, `Sync` and `'static` for `K` ```

Given we might allow users to swap cache implementations eventually, It might indeed make more sense to express these requirements instead.

that would also address the interior mutability concern you voiced, which is adressed by moka

Reply:
I'm going to go ahead and land this as is, I've added a from_cache() function to APQ, which lays out the ground work of several backend implementations, and will give us a bit of time to figure out what kind of traits we're looking for [and adress this comment](https://github.com/apollographql/router/issues/273#issuecomment-1005557523)